pipenv/project.py

```diff
@@ -26,7 +33,13 @@ def __init__(self):
     @property
     def name(self):
         if self._name is None:
-            self._name = self.pipfile_location.split(os.sep)[-2]
+            slug = slugify(self.pipfile_location.split(os.sep)[-2])
+            hash = blake2b(self.pipfile_location.encode(), digest_size=3)
```

Comment:
There's no good way for pipenv to distinguish the next two:

- It's the same project with the different location
- It's the different project with the same name

So we must make a choice in here.

1. Assume that all projects with the same name as same projects. (Current implementation, the "same project name" issue occurs)
2. Assume that all projects with the same name as different projects. (This PR, the "reinstall required after moving project directory" issue occurs)

I thought that the "same project name" issue is more severe than the "reinstall required after moving project directory" issue since the latter one is just an inconvenience while the former one is an issue without a proper way to workaround.

About `virtualenv_location`, good point. I'll move codes at there.

The hash collision probability is 0.03%, and I though that it's practical enough since users won't create ~3000 projects with the same name. Even if it may collide, it's better than the collision with 100% probability without a hash. But I'm OK to use a longer hash if I can sacrifice some beauty. Shall we?

And the hash collision probability has no relation with the name of the virtualenv, and the hash is OK to collide with a project which has a different name. Because it uses `<name>-<hash>` as a key.